In [10]:
import pandas as pd
import openpyxl
import numpy as np
import warnings
import time
warnings.filterwarnings('ignore')

**1.BASE DE DATOS KAGGLE: Importamos la base datos sobre el top 1000 películas de IMDB en junio de 2022**

URL: https://www.kaggle.com/datasets/ramjasmaurya/top-250s-in-imdb

In [13]:
movies=pd.read_csv('data/imdb_movies.csv')

In [14]:
movies.head()

ranking of movie\r\n                                 movie name\r\n   Year  \
0                    1                       The Shawshank Redemption  -1994   
1                    2                                  The Godfather  -1972   
2                    3                                The Dark Knight  -2008   
3                    4  The Lord of the Rings: The Return of the King  -2003   
4                    5                               Schindler's List  -1993   

  certificate  runtime                      genre  RATING  metascore  \
0          15  142 min                      Drama     9.3       81.0   
1           X  175 min               Crime, Drama     9.2      100.0   
2         12A  152 min       Action, Crime, Drama     9.0       84.0   
3         12A  201 min   Action, Adventure, Drama     9.0       94.0   
4          15  195 min  Biography, Drama, History     9.0       94.0   

                                DETAIL ABOUT MOVIE\n          DIRECTOR\r\n  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

        ACTOR 1\n        ACTOR 2\n        ACTOR 3           ACTOR 4  \
0     Tim Robbins   Morgan Freeman     Bob Gunton    William Sadler   
1   Marlon Brando        Al Pacino     James Caan      Diane Keaton   
2  Christian Bale     Heath Ledger  Aaron Eckhart     Michael Caine   
3     Elijah Wood  Viggo Mortensen   Ian McKellen     Orlando Bloom   
4     Liam Neeson    Ralph Fiennes   Ben Kingsley  Caroline Goodall   

       votes GROSS COLLECTION\r\n  
0  2,603,314              $28.34M  
1  1,798,731             $134.97M  
2  2,574,810             $534.86M  
3  1,787,701             $377.85M  
4  1,323,776              $96.90M

Aquí vamos a cambiarle el nombre para que quede más uniforme.

In [15]:
movies.columns=['ranking', 'title', 'year', 'certificate', 'runtime', 'genre', 'rating', 'metascore', 'detail', 'director', 'actor1', 'actor2', 'actor3', 'actor4', 'votes', 'gross_colection']

In [16]:
movies.head()

ranking                                          title   year certificate  \
0       1                       The Shawshank Redemption  -1994          15   
1       2                                  The Godfather  -1972           X   
2       3                                The Dark Knight  -2008         12A   
3       4  The Lord of the Rings: The Return of the King  -2003         12A   
4       5                               Schindler's List  -1993          15   

   runtime                      genre  rating  metascore  \
0  142 min                      Drama     9.3       81.0   
1  175 min               Crime, Drama     9.2      100.0   
2  152 min       Action, Crime, Drama     9.0       84.0   
3  201 min   Action, Adventure, Drama     9.0       94.0   
4  195 min  Biography, Drama, History     9.0       94.0   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

           actor1           actor2         actor3            actor4  \
0     Tim Robbins   Morgan Freeman     Bob Gunton    William Sadler   
1   Marlon Brando        Al Pacino     James Caan      Diane Keaton   
2  Christian Bale     Heath Ledger  Aaron Eckhart     Michael Caine   
3     Elijah Wood  Viggo Mortensen   Ian McKellen     Orlando Bloom   
4     Liam Neeson    Ralph Fiennes   Ben Kingsley  Caroline Goodall   

       votes gross_colection  
0  2,603,314         $28.34M  
1  1,798,731        $134.97M  
2  2,574,810        $534.86M  
3  1,787,701        $377.85M  
4  1,323,776         $96.90M

In [17]:
movies.shape

(1000, 16)

He pasado los datos de title a lista, para poder hacer el scraping que sigue a continuación

In [19]:
peliculas=movies['title'].tolist()

peliculas[:5]

['The Shawshank Redemption',
 'The Godfather',
 'The Dark Knight',
 'The Lord of the Rings: The Return of the King',
 "Schindler's List"]

In [20]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ranking          1000 non-null   object 
 1   title            1000 non-null   object 
 2   year             1000 non-null   object 
 3   certificate      995 non-null    object 
 4   runtime          1000 non-null   object 
 5   genre            1000 non-null   object 
 6   rating           1000 non-null   float64
 7   metascore        837 non-null    float64
 8   detail           1000 non-null   object 
 9   director         1000 non-null   object 
 10  actor1           1000 non-null   object 
 11  actor2           1000 non-null   object 
 12  actor3           1000 non-null   object 
 13  actor4           1000 non-null   object 
 14  votes            1000 non-null   object 
 15  gross_colection  820 non-null    object 
dtypes: float64(2), object(14)
memory usage: 125.1+ KB


In [21]:
#%pip install selenium

In [22]:
from selenium import webdriver

In [23]:
#%pip install webdriver-manager

**2.SCRAPING: Se ha scrapeado la página web de JustWatch, como segunda fuente, para obtener las plataformas disponibles en streaming de cada una de las películas del dataframe.**

In [25]:
from selenium.webdriver.chrome.options import Options


opciones=Options()


opciones.add_experimental_option('excludeSwitches', ['enable-automation'])
opciones.add_experimental_option('useAutomationExtension', False)


opciones.add_argument('user-data-dir=selenium') 

In [26]:
from selenium.webdriver.common.by import By 

import time

import warnings
warnings.filterwarnings('ignore')

In [27]:
from webdriver_manager.chrome import ChromeDriverManager

from selenium.webdriver.common.keys import Keys  

In [28]:
PATH = ChromeDriverManager().install()     
driver=webdriver.Chrome(PATH)     

Con la función que vemos a continuación, he conseguido sacar todas las plataformas en las que están disponibles las películas, obteniendo como resultado una lista, que he añadido al dataframe anterior.

In [29]:

def mi_funcion(x):
    try:
        
        url='https://www.justwatch.com/'
        busqueda = x

        driver=webdriver.Chrome(PATH, options=opciones)

        driver.get(url)

        time.sleep(1.5)

        elemento = driver.find_element(By.XPATH, '//*[@id="app"]/div[3]/div/div[2]/div[1]/div/ion-searchbar/div/input')
        elemento.click()

        time.sleep(1.5)


        elemento.send_keys(busqueda)

        elemento.send_keys(u'\ue007')
        time.sleep(1.5)
        elemento.send_keys(u'\ue007')

        time.sleep(1.5)


        driver.find_element(By.XPATH, '//*[@id="base"]/div[3]/div/div/div[2]/div[1]/div[1]/div[2]/a').click()
        
        time.sleep(1.5)

        a = driver.find_elements(By.XPATH, '//*[@id="base"]/div[2]/div/div[2]/div[2]/div[2]/div[1]/div[3]/div/div[2]/div[2]/div/div/a/picture/img')

        plataformas = []

        for i in a:
            
            plataformas.append(i.get_attribute('title'))

        return plataformas
    except:
        print(f"Error en la búsqueda: {x}")
        return []


In [31]:
from tqdm.notebook import tqdm

In [32]:
list_res = []

for e in tqdm(peliculas):
    try:
        res = mi_funcion(e)
        list_res.append(res)
    except:
        list_res.append("No disponible en streaming")

  0%|          | 0/1000 [00:00<?, ?it/s]

Error en la búsqueda: Rear Window
Error en la búsqueda: M
Error en la búsqueda: Chhichhore
Error en la búsqueda: Z
Error en la búsqueda: Sarfarosh
Error en la búsqueda: Farewell My Concubine
Error en la búsqueda: Sholay
Error en la búsqueda: The Imitation Game
Error en la búsqueda: Mirror
Error en la búsqueda: Young Frankenstein


In [33]:
list_res

[['HBO Max', 'Movistar Plus'],
 ['Netflix', 'Amazon Prime Video', 'HBO Max', 'Netflix basic with Ads'],
 ['HBO Max'],
 ['Amazon Prime Video', 'HBO Max'],
 ['Amazon Prime Video', 'Filmin'],
 ['Amazon Prime Video', 'HBO Max'],
 ['Filmin'],
 ['Amazon Prime Video'],
 ['Netflix', 'Movistar Plus', 'Netflix basic with Ads'],
 ['HBO Max'],
 ['Amazon Prime Video', 'HBO Max'],
 ['Amazon Prime Video', 'Disney Plus', 'Movistar Plus', 'Rakuten TV'],
 ['Amazon Prime Video', 'HBO Max'],
 ['Netflix', 'HBO Max', 'Netflix basic with Ads'],
 ['Amazon Prime Video', 'Movistar Plus', 'FlixOlé'],
 [],
 ['Amazon Video',
  'Apple iTunes',
  'Movistar Plus',
  'Rakuten TV',
  'Google Play Movies',
  'Microsoft Store',
  'Chili'],
 ['HBO Max'],
 ['Disney Plus'],
 ['Amazon Video', 'Apple iTunes', 'Rakuten TV', 'Chili', 'Google Play Movies'],
 ['Amazon Prime Video'],
 ['HBO Max', 'Tivify'],
 ['Amazon Prime Video', 'HBO Max', 'Movistar Plus', 'Tivify', 'Acontra Plus'],
 ['Netflix', 'Netflix basic with Ads'],
 ['Rak

Con la función que vemos a continuación, si hay información sobre la plataforma se añade, y si la lista está vacía, significa que dichas plataformas no estan disponibles

In [34]:
list_of_strings = list(map(lambda x: ', '.join(x) if len(x) > 0 else "No disponible en streaming", list_res))
movies['streaming_services'] = list_of_strings


In [43]:
movies.head()

ranking                                          title   year certificate  \
0       1                       The Shawshank Redemption  -1994          15   
1       2                                  The Godfather  -1972           X   
2       3                                The Dark Knight  -2008         12A   
3       4  The Lord of the Rings: The Return of the King  -2003         12A   
4       5                               Schindler's List  -1993          15   

   runtime                      genre  rating  metascore  \
0  142 min                      Drama     9.3       81.0   
1  175 min               Crime, Drama     9.2      100.0   
2  152 min       Action, Crime, Drama     9.0       84.0   
3  201 min   Action, Adventure, Drama     9.0       94.0   
4  195 min  Biography, Drama, History     9.0       94.0   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

           actor1           actor2         actor3            actor4  \
0     Tim Robbins   Morgan Freeman     Bob Gunton    William Sadler   
1   Marlon Brando        Al Pacino     James Caan      Diane Keaton   
2  Christian Bale     Heath Ledger  Aaron Eckhart     Michael Caine   
3     Elijah Wood  Viggo Mortensen   Ian McKellen     Orlando Bloom   
4     Liam Neeson    Ralph Fiennes   Ben Kingsley  Caroline Goodall   

       votes gross_colection  \
0  2,603,314         $28.34M   
1  1,798,731        $134.97M   
2  2,574,810        $534.86M   
3  1,787,701        $377.85M   
4  1,323,776         $96.90M   

                                  streaming_services  
0                             HBO Max, Movistar Plus  
1  Netflix, Amazon Prime Video, HBO Max, Netflix ...  
2                                            HBO Max  
3                        Amazon Prime Video, HBO Max  
4                         Amazon Prime Video, Filmin

Aquí guardo el dataframe, para no perder los datos.

In [116]:
movies.to_csv('moviesimdb.csv', index=False)

In [117]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ranking             1000 non-null   object 
 1   title               1000 non-null   object 
 2   year                1000 non-null   object 
 3   certificate         1000 non-null   object 
 4   runtime             1000 non-null   object 
 5   genre               1000 non-null   object 
 6   rating              1000 non-null   float64
 7   metascore           837 non-null    float64
 8   detail              1000 non-null   object 
 9   director            1000 non-null   object 
 10  actor1              1000 non-null   object 
 11  actor2              1000 non-null   object 
 12  actor3              1000 non-null   object 
 13  actor4              1000 non-null   object 
 14  votes               1000 non-null   object 
 15  streaming_services  1000 non-null   object 
dtypes: floa

In [121]:
movies=pd.read_csv('data/moviesimdb.csv')

In [122]:
movies.head()

ranking                                          title  year certificate  \
0       1                       The Shawshank Redemption  1994          15   
1       2                                  The Godfather  1972       Adult   
2       3                                The Dark Knight  2008          12   
3       4  The Lord of the Rings: The Return of the King  2003          12   
4       5                               Schindler's List  1993          15   

   runtime                      genre  rating  metascore  \
0  142 min                      Drama     9.3       81.0   
1  175 min               Crime, Drama     9.2      100.0   
2  152 min       Action, Crime, Drama     9.0       84.0   
3  201 min   Action, Adventure, Drama     9.0       94.0   
4  195 min  Biography, Drama, History     9.0       94.0   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

           actor1           actor2         actor3            actor4  \
0     Tim Robbins   Morgan Freeman     Bob Gunton    William Sadler   
1   Marlon Brando        Al Pacino     James Caan      Diane Keaton   
2  Christian Bale     Heath Ledger  Aaron Eckhart     Michael Caine   
3     Elijah Wood  Viggo Mortensen   Ian McKellen     Orlando Bloom   
4     Liam Neeson    Ralph Fiennes   Ben Kingsley  Caroline Goodall   

       votes                                 streaming_services  
0  2,603,314                             HBO Max, Movistar Plus  
1  1,798,731  Netflix, Amazon Prime Video, HBO Max, Netflix ...  
2  2,574,810                                            HBO Max  
3  1,787,701                        Amazon Prime Video, HBO Max  
4  1,323,776                         Amazon Prime Video, Filmin

Limpieza de datos, para que quede todo uniforme.

In [107]:
movies.certificate.unique()
movies['certificate'].replace(['X', '18', 'A', 'AA'], 'Adult', inplace=True)
movies['certificate'].replace(['U', 'All', 'UA'], 'All', inplace=True)
movies['certificate'].replace(['12A', '12'], '12', inplace=True)
movies['certificate'].replace(['UA 13+'], '13', inplace=True)



In [108]:
movies.head()

ranking                                          title   year certificate  \
0       1                       The Shawshank Redemption  -1994          15   
1       2                                  The Godfather  -1972       Adult   
2       3                                The Dark Knight  -2008          12   
3       4  The Lord of the Rings: The Return of the King  -2003          12   
4       5                               Schindler's List  -1993          15   

   runtime                      genre  rating  metascore  \
0  142 min                      Drama     9.3       81.0   
1  175 min               Crime, Drama     9.2      100.0   
2  152 min       Action, Crime, Drama     9.0       84.0   
3  201 min   Action, Adventure, Drama     9.0       94.0   
4  195 min  Biography, Drama, History     9.0       94.0   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

           actor1           actor2         actor3            actor4  \
0     Tim Robbins   Morgan Freeman     Bob Gunton    William Sadler   
1   Marlon Brando        Al Pacino     James Caan      Diane Keaton   
2  Christian Bale     Heath Ledger  Aaron Eckhart     Michael Caine   
3     Elijah Wood  Viggo Mortensen   Ian McKellen     Orlando Bloom   
4     Liam Neeson    Ralph Fiennes   Ben Kingsley  Caroline Goodall   

       votes                                 streaming_services  
0  2,603,314                             HBO Max, Movistar Plus  
1  1,798,731  Netflix, Amazon Prime Video, HBO Max, Netflix ...  
2  2,574,810                                            HBO Max  
3  1,787,701                        Amazon Prime Video, HBO Max  
4  1,323,776                         Amazon Prime Video, Filmin

In [109]:
movies['certificate'].fillna('Not Rated', inplace=True)


In [115]:
movies.year.unique()

array(['1994', '1972', '2008', '2003', '1993', '1974', '1957', '2021',
       '2010', '2002', '1999', '2001', '1966', '2020', '1990', '1980',
       '1975', '2022', '2014', '1998', '1997', '1995', '1991', '1977',
       '1962', '1954', '1946', '2019', '2011', '2006', '2000', '1988',
       '1985', '1979', '1968', '1960', '1942', '1936', '1931', '2018',
       '(I)', '(II)', '2016', '2012', '2009', '1986', '1984', '1981',
       '1963', '1964', '1950', '1940', '2017', '2013', '2007', '2004',
       '1992', '1987', '1983', '1973', '1971', '1961', '1959', '1958',
       '1955', '1952', '1948', '1944', '1941', '1927', '1921', '2015',
       '2005', '1989', '1982', '1976', '1969', '1965', '1953', '1939',
       '1928', '1926', '1925', '1924', '(III)', '1996', '1978', '1967',
       '1951', '1949', '1937', '1934', '1930', '1956', '1947', '1945',
       '1920', '1970', '1943', '1938', '1933', '1932', '1922', '1935'],
      dtype=object)

In [123]:
movies.head()

ranking                                          title  year certificate  \
0       1                       The Shawshank Redemption  1994          15   
1       2                                  The Godfather  1972       Adult   
2       3                                The Dark Knight  2008          12   
3       4  The Lord of the Rings: The Return of the King  2003          12   
4       5                               Schindler's List  1993          15   

   runtime                      genre  rating  metascore  \
0  142 min                      Drama     9.3       81.0   
1  175 min               Crime, Drama     9.2      100.0   
2  152 min       Action, Crime, Drama     9.0       84.0   
3  201 min   Action, Adventure, Drama     9.0       94.0   
4  195 min  Biography, Drama, History     9.0       94.0   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

           actor1           actor2         actor3            actor4  \
0     Tim Robbins   Morgan Freeman     Bob Gunton    William Sadler   
1   Marlon Brando        Al Pacino     James Caan      Diane Keaton   
2  Christian Bale     Heath Ledger  Aaron Eckhart     Michael Caine   
3     Elijah Wood  Viggo Mortensen   Ian McKellen     Orlando Bloom   
4     Liam Neeson    Ralph Fiennes   Ben Kingsley  Caroline Goodall   

       votes                                 streaming_services  
0  2,603,314                             HBO Max, Movistar Plus  
1  1,798,731  Netflix, Amazon Prime Video, HBO Max, Netflix ...  
2  2,574,810                                            HBO Max  
3  1,787,701                        Amazon Prime Video, HBO Max  
4  1,323,776                         Amazon Prime Video, Filmin

He hecho una copia del dataframe, ya que al limpiar la columna year, tuve problemas y preferí asegurarme de no perder los datos originales.

In [124]:
moviess=movies.copy()

In [127]:
moviess.head()

ranking                                          title  year certificate  \
0       1                       The Shawshank Redemption  1994          15   
1       2                                  The Godfather  1972       Adult   
2       3                                The Dark Knight  2008          12   
3       4  The Lord of the Rings: The Return of the King  2003          12   
4       5                               Schindler's List  1993          15   

   runtime                      genre  rating  metascore  \
0  142 min                      Drama     9.3       81.0   
1  175 min               Crime, Drama     9.2      100.0   
2  152 min       Action, Crime, Drama     9.0       84.0   
3  201 min   Action, Adventure, Drama     9.0       94.0   
4  195 min  Biography, Drama, History     9.0       94.0   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

           actor1           actor2         actor3            actor4  \
0     Tim Robbins   Morgan Freeman     Bob Gunton    William Sadler   
1   Marlon Brando        Al Pacino     James Caan      Diane Keaton   
2  Christian Bale     Heath Ledger  Aaron Eckhart     Michael Caine   
3     Elijah Wood  Viggo Mortensen   Ian McKellen     Orlando Bloom   
4     Liam Neeson    Ralph Fiennes   Ben Kingsley  Caroline Goodall   

       votes                                 streaming_services  
0  2,603,314                             HBO Max, Movistar Plus  
1  1,798,731  Netflix, Amazon Prime Video, HBO Max, Netflix ...  
2  2,574,810                                            HBO Max  
3  1,787,701                        Amazon Prime Video, HBO Max  
4  1,323,776                         Amazon Prime Video, Filmin

In [130]:
moviess.year.unique()

array(['1994', '1972', '2008', '2003', '1993', '1974', '1957', '2021',
       '2010', '2002', '1999', '2001', '1966', '2020', '1990', '1980',
       '1975', '2022', '2014', '1998', '1997', '1995', '1991', '1977',
       '1962', '1954', '1946', '2019', '2011', '2006', '2000', '1988',
       '1985', '1979', '1968', '1960', '1942', '1936', '1931', '2018',
       '(I)', '(II)', '2016', '2012', '2009', '1986', '1984', '1981',
       '1963', '1964', '1950', '1940', '2017', '2013', '2007', '2004',
       '1992', '1987', '1983', '1973', '1971', '1961', '1959', '1958',
       '1955', '1952', '1948', '1944', '1941', '1927', '1921', '2015',
       '2005', '1989', '1982', '1976', '1969', '1965', '1953', '1939',
       '1928', '1926', '1925', '1924', '(III)', '1996', '1978', '1967',
       '1951', '1949', '1937', '1934', '1930', '1956', '1947', '1945',
       '1920', '1970', '1943', '1938', '1933', '1932', '1922', '1935'],
      dtype=object)

In [133]:
movies=pd.read_csv('data/imdb_movies.csv')

In [198]:
movies.Year.unique()

array(['-1994', '-1972', '-2008', '-2003', '-1993', '-1974', '-1957',
       '-2021', '-2010', '-2002', '-1999', '-2001', '-1966', '-2020',
       '-1990', '-1980', '-1975', '-2022', '-2014', '-1998', '-1997',
       '-1995', '-1991', '-1977', '-1962', '-1954', '-1946', '-2019',
       '-2011', '-2006', '-2000', '-1988', '-1985', '-1979', '-1968',
       '-1960', '-1942', '-1936', '-1931', '-2018', '(I) (2019)',
       '(II) (2018)', '-2016', '(I) (2017)', '-2012', '-2009', '-1986',
       '-1984', '-1981', '-1963', '-1964', '-1950', '-1940', '-2017',
       '-2013', '-2007', '-2004', '-1992', '-1987', '-1983', '-1973',
       '-1971', '-1961', '-1959', '-1958', '-1955', '-1952', '-1948',
       '-1944', '-1941', '-1927', '-1921', '-2015', '(I) (2015)',
       '(I) (2020)', '-2005', '-1989', '-1982', '-1976', '-1969', '-1965',
       '-1953', '-1939', '-1928', '-1926', '-1925', '-1924',
       '(III) (2016)', '(I) (2013)', '-1996', '-1978', '-1967', '-1951',
       '-1949', '-1937', '-

In [199]:
moviess['year'] = movies['Year']


In [200]:
moviess.year.unique()

array(['-1994', '-1972', '-2008', '-2003', '-1993', '-1974', '-1957',
       '-2021', '-2010', '-2002', '-1999', '-2001', '-1966', '-2020',
       '-1990', '-1980', '-1975', '-2022', '-2014', '-1998', '-1997',
       '-1995', '-1991', '-1977', '-1962', '-1954', '-1946', '-2019',
       '-2011', '-2006', '-2000', '-1988', '-1985', '-1979', '-1968',
       '-1960', '-1942', '-1936', '-1931', '-2018', '(I) (2019)',
       '(II) (2018)', '-2016', '(I) (2017)', '-2012', '-2009', '-1986',
       '-1984', '-1981', '-1963', '-1964', '-1950', '-1940', '-2017',
       '-2013', '-2007', '-2004', '-1992', '-1987', '-1983', '-1973',
       '-1971', '-1961', '-1959', '-1958', '-1955', '-1952', '-1948',
       '-1944', '-1941', '-1927', '-1921', '-2015', '(I) (2015)',
       '(I) (2020)', '-2005', '-1989', '-1982', '-1976', '-1969', '-1965',
       '-1953', '-1939', '-1928', '-1926', '-1925', '-1924',
       '(III) (2016)', '(I) (2013)', '-1996', '-1978', '-1967', '-1951',
       '-1949', '-1937', '-

In [201]:
moviess['year']=moviess['year'].str.replace('-', '')

In [202]:
moviess.year.unique()

array(['1994', '1972', '2008', '2003', '1993', '1974', '1957', '2021',
       '2010', '2002', '1999', '2001', '1966', '2020', '1990', '1980',
       '1975', '2022', '2014', '1998', '1997', '1995', '1991', '1977',
       '1962', '1954', '1946', '2019', '2011', '2006', '2000', '1988',
       '1985', '1979', '1968', '1960', '1942', '1936', '1931', '2018',
       '(I) (2019)', '(II) (2018)', '2016', '(I) (2017)', '2012', '2009',
       '1986', '1984', '1981', '1963', '1964', '1950', '1940', '2017',
       '2013', '2007', '2004', '1992', '1987', '1983', '1973', '1971',
       '1961', '1959', '1958', '1955', '1952', '1948', '1944', '1941',
       '1927', '1921', '2015', '(I) (2015)', '(I) (2020)', '2005', '1989',
       '1982', '1976', '1969', '1965', '1953', '1939', '1928', '1926',
       '1925', '1924', '(III) (2016)', '(I) (2013)', '1996', '1978',
       '1967', '1951', '1949', '1937', '1934', '1930', '(I) (2014)',
       '(I) (2004)', '1956', '1947', '1945', '1920', '(II) (2016)',
      

In [203]:
list_mov= moviess.year.tolist()



In [214]:
new_years = []
for year in list_mov:
    match = re.search(r'\((?:[IVXLCDM]+\s)?(\d+)\)', year)
    if match:
        new_years.append(match.group(1))
    else:
        new_years.append(year)

new_years[:5]


['1994', '1972', '2008', '2003', '1993']

In [206]:
moviess['year'] = new_years


In [207]:
moviess.year.unique()

array(['1994', '1972', '2008', '2003', '1993', '1974', '1957', '2021',
       '2010', '2002', '1999', '2001', '1966', '2020', '1990', '1980',
       '1975', '2022', '2014', '1998', '1997', '1995', '1991', '1977',
       '1962', '1954', '1946', '2019', '2011', '2006', '2000', '1988',
       '1985', '1979', '1968', '1960', '1942', '1936', '1931', '2018',
       '2016', '2017', '2012', '2009', '1986', '1984', '1981', '1963',
       '1964', '1950', '1940', '2013', '2007', '2004', '1992', '1987',
       '1983', '1973', '1971', '1961', '1959', '1958', '1955', '1952',
       '1948', '1944', '1941', '1927', '1921', '2015', '2005', '1989',
       '1982', '1976', '1969', '1965', '1953', '1939', '1928', '1926',
       '1925', '1924', '1996', '1978', '1967', '1951', '1949', '1937',
       '1934', '1930', '1956', '1947', '1945', '1920', '1970', '1943',
       '1938', '1933', '1932', '1922', '1935'], dtype=object)

In [216]:
moviess.year.unique()

array(['1994', '1972', '2008', '2003', '1993', '1974', '1957', '2021',
       '2010', '2002', '1999', '2001', '1966', '2020', '1990', '1980',
       '1975', '2022', '2014', '1998', '1997', '1995', '1991', '1977',
       '1962', '1954', '1946', '2019', '2011', '2006', '2000', '1988',
       '1985', '1979', '1968', '1960', '1942', '1936', '1931', '2018',
       '2016', '2017', '2012', '2009', '1986', '1984', '1981', '1963',
       '1964', '1950', '1940', '2013', '2007', '2004', '1992', '1987',
       '1983', '1973', '1971', '1961', '1959', '1958', '1955', '1952',
       '1948', '1944', '1941', '1927', '1921', '2015', '2005', '1989',
       '1982', '1976', '1969', '1965', '1953', '1939', '1928', '1926',
       '1925', '1924', '1996', '1978', '1967', '1951', '1949', '1937',
       '1934', '1930', '1956', '1947', '1945', '1920', '1970', '1943',
       '1938', '1933', '1932', '1922', '1935'], dtype=object)

In [217]:
moviess['year'] = moviess['year'].astype(int)


In [219]:
moviess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ranking             1000 non-null   object 
 1   title               1000 non-null   object 
 2   year                1000 non-null   int64  
 3   certificate         1000 non-null   object 
 4   runtime             1000 non-null   object 
 5   genre               1000 non-null   object 
 6   rating              1000 non-null   float64
 7   metascore           837 non-null    float64
 8   detail              1000 non-null   object 
 9   director            1000 non-null   object 
 10  actor1              1000 non-null   object 
 11  actor2              1000 non-null   object 
 12  actor3              1000 non-null   object 
 13  actor4              1000 non-null   object 
 14  votes               1000 non-null   object 
 15  streaming_services  1000 non-null   object 
dtypes: floa

In [224]:
movie=moviess.copy()

In [252]:
movie.head()

ranking                                          title  year certificate  \
0       1                       The Shawshank Redemption  1994          15   
1       2                                  The Godfather  1972       Adult   
2       3                                The Dark Knight  2008          12   
3       4  The Lord of the Rings: The Return of the King  2003          12   
4       5                               Schindler's List  1993          15   

   runtime                      genre  rating  metascore  \
0  142 min                      Drama     9.3       81.0   
1  175 min               Crime, Drama     9.2      100.0   
2  152 min       Action, Crime, Drama     9.0       84.0   
3  201 min   Action, Adventure, Drama     9.0       94.0   
4  195 min  Biography, Drama, History     9.0       94.0   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

           actor1           actor2         actor3            actor4  \
0     Tim Robbins   Morgan Freeman     Bob Gunton    William Sadler   
1   Marlon Brando        Al Pacino     James Caan      Diane Keaton   
2  Christian Bale     Heath Ledger  Aaron Eckhart     Michael Caine   
3     Elijah Wood  Viggo Mortensen   Ian McKellen     Orlando Bloom   
4     Liam Neeson    Ralph Fiennes   Ben Kingsley  Caroline Goodall   

       votes                                 streaming_services  \
0  2,603,314                             HBO Max, Movistar Plus   
1  1,798,731  Netflix, Amazon Prime Video, HBO Max, Netflix ...   
2  2,574,810                                            HBO Max   
3  1,787,701                        Amazon Prime Video, HBO Max   
4  1,323,776                         Amazon Prime Video, Filmin   

                                     title_clean  
0                       the shawshank redemption  
1                                  the godfather  
2                                the dark knight  
3  the lord of the rings: the return of the king  
4                               schindler's list

A conitnuación, voy a eliminar la columna metascrore, ya que proximamente haré una comparativa de estas peliculas, con las notas que tienen en filmaffinitty.

In [268]:
movie = movie.drop("metascore", axis=1)

In [270]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ranking             1000 non-null   object 
 1   title               1000 non-null   object 
 2   year                1000 non-null   int64  
 3   certificate         1000 non-null   object 
 4   runtime             1000 non-null   object 
 5   genre               1000 non-null   object 
 6   rating              1000 non-null   float64
 7   detail              1000 non-null   object 
 8   director            1000 non-null   object 
 9   actor1              1000 non-null   object 
 10  actor2              1000 non-null   object 
 11  actor3              1000 non-null   object 
 12  actor4              1000 non-null   object 
 13  votes               1000 non-null   object 
 14  streaming_services  1000 non-null   object 
 15  title_clean         1000 non-null   object 
dtypes: floa

In [271]:
movie.to_csv('moviesimdb2.csv', index=False) #Este sería el csv final, limpio y preparado.

**3.BASE DE DATOS ZENODO: Para la tercera fuente de información, se ha importado un csv de zenodo con 119003 películas de filmaffinitty.**

URL: https://zenodo.org/record/4249401#.Y-f6A9LMJkg

In [253]:
film=pd.read_csv('/data/filmaf.csv')

In [272]:
film.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119003 entries, 0 to 119002
Data columns (total 9 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  119003 non-null  int64 
 1   Título      119003 non-null  object
 2   Año         119003 non-null  int64 
 3   País        119003 non-null  object
 4   Dirección   118202 non-null  object
 5   Reparto     94525 non-null   object
 6   Nota        119003 non-null  object
 7   Tipo filme  119003 non-null  object
 8   Género      119003 non-null  object
dtypes: int64(2), object(7)
memory usage: 8.2+ MB


In [273]:
film.head()

Unnamed: 0                                Título   Año            País  \
0           0                              '49-'17   1917  Estados Unidos   
1           1                 10,000 Years B.C. (C)  1916  Estados Unidos   
2           2                                 1812   1912           Rusia   
3           3  20.000 leguas de viaje submarino (C)  1907         Francia   
4           4                        A Bad Case (C)  1909         Francia   

                                         Dirección  \
0                                 Ruth Ann Baldwin   
1                                Willis H. O'Brien   
2  Vasili Goncharov, Kai Hansen, Aleksandr Uralsky   
3                                   Georges Méliès   
4                                       Émile Cohl   

                                             Reparto Nota    Tipo filme  \
0  Joseph W. Girard, Leo Pierson, William Dyer, M...   --      Película   
1                                                NaN  5,1  Cortometraje   
2  Pavel Knorr, Vasili Goncharov, Aleksandra Gonc...   --      Película   
3                                     Georges Méliès  6,0  Cortometraje   
4                                                NaN  5,3  Cortometraje   

       Género  
0     Western  
1     Comedia  
2       Drama  
3  Fantástico  
4     Comedia

A continuación he decidido unir ambas bases de datos, para obtener finalmente una única tabla, comparando la nota de filmaffinitty y la de IMDB, e incluir tambien el país de origen de la pelicula.
Para ello, he tenido que seguir los siguientes pasos:

1. Voy a renombrar las columnas, para que coincidan perfectamente en ambos dataframes.

In [274]:
film.rename(columns={'Dirección': 'director'}, inplace=True)
film.rename(columns={'Año': 'year'}, inplace=True)


In [275]:
film.head()

Unnamed: 0                                Título  year            País  \
0           0                              '49-'17   1917  Estados Unidos   
1           1                 10,000 Years B.C. (C)  1916  Estados Unidos   
2           2                                 1812   1912           Rusia   
3           3  20.000 leguas de viaje submarino (C)  1907         Francia   
4           4                        A Bad Case (C)  1909         Francia   

                                          director  \
0                                 Ruth Ann Baldwin   
1                                Willis H. O'Brien   
2  Vasili Goncharov, Kai Hansen, Aleksandr Uralsky   
3                                   Georges Méliès   
4                                       Émile Cohl   

                                             Reparto Nota    Tipo filme  \
0  Joseph W. Girard, Leo Pierson, William Dyer, M...   --      Película   
1                                                NaN  5,1  Cortometraje   
2  Pavel Knorr, Vasili Goncharov, Aleksandra Gonc...   --      Película   
3                                     Georges Méliès  6,0  Cortometraje   
4                                                NaN  5,3  Cortometraje   

       Género  
0     Western  
1     Comedia  
2       Drama  
3  Fantástico  
4     Comedia

In [276]:
film.rename(columns={'País': 'country'}, inplace=True)


2. A continuación voy a mergear ambos dataframes, por las columnas director y año, ya que es donde puede haber una coincidencia real.


In [277]:
merged_df = pd.merge(movie, film, on=['director', 'year'])

In [279]:
merged_df.head()

ranking                                          title  year certificate  \
0       1                       The Shawshank Redemption  1994          15   
1       2                                  The Godfather  1972       Adult   
2       3                                The Dark Knight  2008          12   
3       4  The Lord of the Rings: The Return of the King  2003          12   
4       5                               Schindler's List  1993          15   

   runtime                      genre  rating  \
0  142 min                      Drama     9.3   
1  175 min               Crime, Drama     9.2   
2  152 min       Action, Crime, Drama     9.0   
3  201 min   Action, Adventure, Drama     9.0   
4  195 min  Biography, Drama, History     9.0   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

           actor1  ...      votes  \
0     Tim Robbins  ...  2,603,314   
1   Marlon Brando  ...  1,798,731   
2  Christian Bale  ...  2,574,810   
3     Elijah Wood  ...  1,787,701   
4     Liam Neeson  ...  1,323,776   

                                  streaming_services  \
0                             HBO Max, Movistar Plus   
1  Netflix, Amazon Prime Video, HBO Max, Netflix ...   
2                                            HBO Max   
3                        Amazon Prime Video, HBO Max   
4                         Amazon Prime Video, Filmin   

                                     title_clean Unnamed: 0  \
0                       the shawshank redemption      47226   
1                                  the godfather      28878   
2                                the dark knight      76097   
3  the lord of the rings: the return of the king      77054   
4                               schindler's list      57631   

                                         Título         country  \
0                              Cadena perpetua   Estados Unidos   
1                                   El padrino   Estados Unidos   
2                          El caballero oscuro   Estados Unidos   
3  El señor de los anillos: El retorno del rey    Nueva Zelanda   
4                        La lista de Schindler   Estados Unidos   

                                             Reparto Nota Tipo filme  \
0  Tim Robbins, Morgan Freeman, Bob Gunton, James...  8,6   Película   
1  Marlon Brando, Al Pacino, James Caan, Robert D...  9,0   Película   
2  Christian Bale, Heath Ledger, Aaron Eckhart, M...  8,1   Película   
3  Elijah Wood, Viggo Mortensen, Ian McKellen, Se...  8,2   Película   
4  Liam Neeson, Ben Kingsley, Ralph Fiennes, Caro...  8,6   Película   

                          Género  
0                          Drama  
1                          Drama  
2        Acción, Drama, Thriller  
3  Acción, Aventuras, Fantástico  
4                          Drama  

[5 rows x 23 columns]

Al investigar esta tabla, me dí cuenta que habia varias peliculas, que coincidian tanto en director como en año, por tanto tuve que filtrar por actores y reparto para que me diera una comparación más exacta.

In [287]:
filtro = []
for index, row in merged_df.iterrows():
    
    try:
        if row["actor1"] in row["Reparto"] and (row["actor2"] in row["Reparto"]):
            filtro.append(row)
        
    except:
        continue
top_movies= pd.DataFrame(filtro)

In [288]:
top_movies.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 741 entries, 0 to 1022
Data columns (total 23 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ranking             741 non-null    object 
 1   title               741 non-null    object 
 2   year                741 non-null    int64  
 3   certificate         741 non-null    object 
 4   runtime             741 non-null    object 
 5   genre               741 non-null    object 
 6   rating              741 non-null    float64
 7   detail              741 non-null    object 
 8   director            741 non-null    object 
 9   actor1              741 non-null    object 
 10  actor2              741 non-null    object 
 11  actor3              741 non-null    object 
 12  actor4              741 non-null    object 
 13  votes               741 non-null    object 
 14  streaming_services  741 non-null    object 
 15  title_clean         741 non-null    object 
 16  Unnamed

Aquí me di cuenta que si ponía solo dos actores, era posible que no fuera muy exacta la comparación, y saberlo era complicado, ya que existía la posbilidad de ambos actores estuviesen en la misma pelicula varias veces, por tanto decidí añadir los 4 actores en la comparación, y el resultado fue el siguiente: 

In [293]:
filtered_rows = []
for index, row in merged_df.iterrows():
    
    try:
        if row["actor1"] in row["Reparto"] and (row["actor2"] in row["Reparto"]) and (row["actor3"] in row["Reparto"]) and (row["actor4"] in row["Reparto"]):
            filtered_rows.append(row)
        
    except:
        continue
movies_final= pd.DataFrame(filtered_rows)

In [294]:
movies_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 618 entries, 0 to 1022
Data columns (total 23 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ranking             618 non-null    object 
 1   title               618 non-null    object 
 2   year                618 non-null    int64  
 3   certificate         618 non-null    object 
 4   runtime             618 non-null    object 
 5   genre               618 non-null    object 
 6   rating              618 non-null    float64
 7   detail              618 non-null    object 
 8   director            618 non-null    object 
 9   actor1              618 non-null    object 
 10  actor2              618 non-null    object 
 11  actor3              618 non-null    object 
 12  actor4              618 non-null    object 
 13  votes               618 non-null    object 
 14  streaming_services  618 non-null    object 
 15  title_clean         618 non-null    object 
 16  Unnamed

Finalmente, de 1000 peliculas decidí quedarme con 618, y saber con precisión que la comparación de ambos dataframes era exacta.
A continuación, procedo a la limpieza de la tabla final.

In [297]:
movies_final = movies_final.rename(columns={'Título': 'spanish_title', 'Reparto': 'Actors', 'rating': 'rating_imdb', 'Nota': 'rating_filmaffinitty'})



In [316]:
mov= movies_final[['ranking','title', 'spanish_title', 'year', 'genre', 'certificate', 'detail','director','Actors','country', 'rating_imdb','rating_filmaffinitty', 'streaming_services' ]]

In [318]:
mov = mov.rename(columns={'Actors': 'actors'})

In [319]:
mov.head()

ranking                                          title  \
0       1                       The Shawshank Redemption   
1       2                                  The Godfather   
2       3                                The Dark Knight   
3       4  The Lord of the Rings: The Return of the King   
4       5                               Schindler's List   

                                  spanish_title  year  \
0                              Cadena perpetua   1994   
1                                   El padrino   1972   
2                          El caballero oscuro   2008   
3  El señor de los anillos: El retorno del rey   2003   
4                        La lista de Schindler   1993   

                       genre certificate  \
0                      Drama          15   
1               Crime, Drama       Adult   
2       Action, Crime, Drama          12   
3   Action, Adventure, Drama          12   
4  Biography, Drama, History          15   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

                                              actors         country  \
0  Tim Robbins, Morgan Freeman, Bob Gunton, James...  Estados Unidos   
1  Marlon Brando, Al Pacino, James Caan, Robert D...  Estados Unidos   
2  Christian Bale, Heath Ledger, Aaron Eckhart, M...  Estados Unidos   
3  Elijah Wood, Viggo Mortensen, Ian McKellen, Se...   Nueva Zelanda   
4  Liam Neeson, Ben Kingsley, Ralph Fiennes, Caro...  Estados Unidos   

   rating_imdb rating_filmaffinitty  \
0          9.3                  8,6   
1          9.2                  9,0   
2          9.0                  8,1   
3          9.0                  8,2   
4          9.0                  8,6   

                                  streaming_services  
0                             HBO Max, Movistar Plus  
1  Netflix, Amazon Prime Video, HBO Max, Netflix ...  
2                                            HBO Max  
3                        Amazon Prime Video, HBO Max  
4                         Amazon Prime Video, Filmin

In [306]:
mov.country.unique()

array(['Estados Unidos', 'Nueva Zelanda', 'Italia', 'Japón',
       'Reino Unido', 'Francia', 'Alemania', 'India', 'Dinamarca',
       'Argentina', 'Turquía', 'España', 'Corea del Sur', 'Estonia',
       'del Oeste (RFA)', 'Suecia', 'Australia', 'Brasil', 'México',
       'Irlanda', 'Indonesia', 'Austria', 'Rumanía', 'Canadá',
       'a y Herzegovina', 'Irán', 'Bajos (Holanda)', 'Rusia'],
      dtype=object)

Vamos a reemplazar los países que están en español, por sus abreviaturas en inglés para tener un dataframe uniforme.

In [320]:
replace_c = {
    'Estados Unidos': 'USA',
    'Nueva Zelanda': 'NZ',
    'Italia': 'IT',
    'Japón': 'JP',
    'Reino Unido': 'UK',
    'Francia': 'FR',
    'Alemania': 'DE',
    'India': 'IN',
    'Dinamarca': 'DK',
    'Argentina': 'AR',
    'Turquía': 'TR',
    'España': 'ES',
    'Corea del Sur': 'KR',
    'Estonia': 'EE',
    'del Oeste (RFA)': 'West Germany',
    'Suecia': 'SE',
    'Australia': 'AU',
    'Brasil': 'BR',
    'México': 'MX',
    'Irlanda': 'IE',
    'Indonesia': 'ID',
    'Austria': 'AT',
    'Rumanía': 'RO',
    'Canadá': 'CA',
    'a y Herzegovina': 'BA',
    'Irán': 'IR',
    'Bajos (Holanda)': 'NL',
    'Rusia': 'RU'
}

mov['country'] = mov['country'].replace(replace_c)

In [321]:
mov.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 618 entries, 0 to 1022
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   ranking               618 non-null    object 
 1   title                 618 non-null    object 
 2   spanish_title         618 non-null    object 
 3   year                  618 non-null    int64  
 4   genre                 618 non-null    object 
 5   certificate           618 non-null    object 
 6   detail                618 non-null    object 
 7   director              618 non-null    object 
 8   actors                618 non-null    object 
 9   country               618 non-null    object 
 10  rating_imdb           618 non-null    float64
 11  rating_filmaffinitty  618 non-null    object 
 12  streaming_services    618 non-null    object 
dtypes: float64(1), int64(1), object(11)
memory usage: 67.6+ KB


Reseteamos el ranking según nota de IMDB, ya que se han eliminado algunas filas.

In [326]:
mov = mov.sort_values(by='rating_imdb', ascending=False)


In [332]:
mov = mov.reset_index(drop=True)
mov['ranking'] = range(1, len(df) + 1)


In [334]:
mov.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 618 entries, 0 to 617
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   ranking               618 non-null    int64  
 1   title                 618 non-null    object 
 2   spanish_title         618 non-null    object 
 3   year                  618 non-null    int64  
 4   genre                 618 non-null    object 
 5   certificate           618 non-null    object 
 6   detail                618 non-null    object 
 7   director              618 non-null    object 
 8   actors                618 non-null    object 
 9   country               618 non-null    object 
 10  rating_imdb           618 non-null    float64
 11  rating_filmaffinitty  618 non-null    object 
 12  streaming_services    618 non-null    object 
dtypes: float64(1), int64(2), object(10)
memory usage: 62.9+ KB


Intenté convertir la columna rating_filmaffinitty en float también, pero no es posible ya que algunas peliculas no tienen ninguna nota.

In [339]:
mov.rating_filmaffinitty.unique()

array(['8,6', '9,0', '8,1', '8,2', '8,9', '8,7', '8,0', '8,4', '7,8',
       '7,9', '8,5', '7,5', '7,6', '8,3', '7,4', '7,7', '7,3', '7,0',
       '7,2', '6,9', '6,5', '7,1', '6,1', '6,4', '--', '6,2', '6,8',
       '6,7', '6,6', '5,7', '6,3', '6,0', '5,4', '5,9'], dtype=object)

In [340]:
mov['rating_filmaffinitty'].replace('--', 'NR', inplace=True)

In [341]:
mov.rating_filmaffinitty.unique()

array(['8,6', '9,0', '8,1', '8,2', '8,9', '8,7', '8,0', '8,4', '7,8',
       '7,9', '8,5', '7,5', '7,6', '8,3', '7,4', '7,7', '7,3', '7,0',
       '7,2', '6,9', '6,5', '7,1', '6,1', '6,4', 'NR', '6,2', '6,8',
       '6,7', '6,6', '5,7', '6,3', '6,0', '5,4', '5,9'], dtype=object)

In [349]:
mov['rating_filmaffinitty'] = mov['rating_filmaffinitty'].str.replace(',', '.')

In [350]:
mov.head()

ranking                                          title  \
0        1                       The Shawshank Redemption   
1        2                                  The Godfather   
2        3                                The Dark Knight   
3        4  The Lord of the Rings: The Return of the King   
4        5                               Schindler's List   

                                  spanish_title  year  \
0                              Cadena perpetua   1994   
1                                   El padrino   1972   
2                          El caballero oscuro   2008   
3  El señor de los anillos: El retorno del rey   2003   
4                        La lista de Schindler   1993   

                       genre certificate  \
0                      Drama          15   
1               Crime, Drama       Adult   
2       Action, Crime, Drama          12   
3   Action, Adventure, Drama          12   
4  Biography, Drama, History          15   

                                              detail              director  \
0  Two imprisoned men bond over a number of years...        Frank Darabont   
1  The aging patriarch of an organized crime dyna...  Francis Ford Coppola   
2  When the menace known as the Joker wreaks havo...     Christopher Nolan   
3  Gandalf and Aragorn lead the World of Men agai...         Peter Jackson   
4  In German-occupied Poland during World War II,...      Steven Spielberg   

                                              actors country  rating_imdb  \
0  Tim Robbins, Morgan Freeman, Bob Gunton, James...     USA          9.3   
1  Marlon Brando, Al Pacino, James Caan, Robert D...     USA          9.2   
2  Christian Bale, Heath Ledger, Aaron Eckhart, M...     USA          9.0   
3  Elijah Wood, Viggo Mortensen, Ian McKellen, Se...      NZ          9.0   
4  Liam Neeson, Ben Kingsley, Ralph Fiennes, Caro...     USA          9.0   

  rating_filmaffinitty                                 streaming_services  
0                  8.6                             HBO Max, Movistar Plus  
1                  9.0  Netflix, Amazon Prime Video, HBO Max, Netflix ...  
2                  8.1                                            HBO Max  
3                  8.2                        Amazon Prime Video, HBO Max  
4                  8.6                         Amazon Prime Video, Filmin

**4. Exportamos la tabla a la base de datos movies en SQL**

In [355]:
from sqlalchemy import create_engine

str_conn = 'mysql+pymysql://root:3164@localhost:3306'

cursor = create_engine(str_conn)

cursor.execute('drop database if exists movies;')

cursor.execute('create database movies;')

In [356]:
str_conn = 'mysql+pymysql://root:3164@localhost:3306/movies'

cursor = create_engine(str_conn)

In [357]:
mov.to_sql(name='top_movies', con=cursor, if_exists='replace', index=False)

618